### PR TITLE
Add support for scraping grouped documents to the MendeleyScraper

### LINKED ISSUE #22: Add a new Mendeley scraper (groups)
The Mendeley scraper class implemented in #21 should be modified as follows:
- The class should have a method to scrape documents inside groups identified by group ids. If no max is given, it exhausts all documents in specified groups. If no group ids are given, the ungrouped documents should be retrieved instead.
- The API endpoint can be found [here](https://dev.mendeley.com/methods/#retrieving-documents).

### PR DESCRIPTION
**Pull request by kohrVid**
Closes #22

This PR refactors the `MendeleyScraper` class, allowing users to scrape the Mendeley Reference Manager by group ID(s).

Example usage would be:

    articles = []
    access_token = "ACCESS_TOKEN"
    collection_type = "folders"
    collection_ids = [folder_id_1, folder_id_2]

    scraper = RayyanScrapers::MendeleyScraper.new(access_token)
    scraper.scrape(collection_type, collection_ids) do |article, total|
      articles << { total: total, article: article }
    end

Grouped documents can be filtered using the `groups` collection type:

    articles = []
    access_token = "ACCESS_TOKEN"
    collection_type = "groups"
    collection_ids = [group_id_1, group_id_2]

    scraper = RayyanScrapers::MendeleyScraper.new(access_token)
    scraper.scrape(collection_type, collection_ids) do |article, total|
      articles << { total: total, article: article }
    end

Unfiltered lists can be returned using a blank collection type (e.g., `nil` or
`''`) without a list of collection IDs:

    articles = []
    access_token = "ACCESS_TOKEN"
    collection_type = ""

    scraper = RayyanScrapers::MendeleyScraper.new(access_token)
    scraper.scrape(collection_type) do |article, total|
      articles << { total: total, article: article }
    end

An unknown collection type will raise will raise an `UnknownCollectionType` error.

### CODE REVIEW

File: spec/mendeley_scraper_spec.rb [hunk 1]
```diff
@@ -397,17 +408,38 @@
...
+          'https://api.mendeley.com/documents?group_id=2'
+        ])
+      end
+
+      context 'when collection_ids is empty' do
+        let(:collection_ids) { [] }
+
+        it 'returns a list of search URLs for the groups' do
+          expect(mendeley_scraper.send(:search_urls, collection_type, collection_ids)).to match_array([
+            'https://api.mendeley.com/documents?group_id='
```
Comment (hammady) [inline spec/mendeley_scraper_spec.rb:437]: Can this query string parameter be omitted?

**Review by hammady — COMMENTED** [review]
From quickly reading the API and my demo back then, I remember fetching groups does not require further fetching for articles as everything is already contained in the list pages. Does this PR assume so? I only observed changes in the search URL creation but everything else remains the same, where fetching folders require fetching articles. Please confirm.

And by the way, I found the *real* reason `blank?` was not failing, here it is: https://github.com/rayyansys/rayyan-scrapers/blob/master/lib/support/blank.rb

File: README.md [hunk 2]
```diff
@@ -116,6 +116,22 @@ be parsed using code like the following:
...
+    access_token = "ACCESS_TOKEN"
+    collection_type = "groups"
+    collection_ids = [group_id_1, group_id_2]
+
+    scraper = RayyanScrapers::MendeleyScraper.new(access_token)
+    scraper.scrape(collection_type, collection_ids) do |article, total|
+      articles << { total: total, article: article }
+    end
+
+Note, if the `#scrape` method is called with the `groups` collection type and
```
[the remark this answers is not in the document]
Comment (kohrVid) [reply README.md:131]: I think I've gotten this working now though I'm a little unsure about the third requirement. Is something like my latest commit what you had in mind?

File: spec/mendeley_scraper_spec.rb (hunk 1 above)
Comment (kohrVid) [reply spec/mendeley_scraper_spec.rb]: This can be done. I'd previously thought that adding the parameter would return all documents without a group assigned but it doesn't seem like this functionality exists in the Mendeley API. Providing the parameter with a nil value returns a list of documents and ignores the grouping altogether which is essentially what the `/documents` endpoint does without a query string

File: lib/rayyan-scrapers/mendeley_scraper.rb [hunk 3]
```diff
@@ -30,6 +30,8 @@ def self.results_per_page
     end
 
     def iterate_list_pages(page, &block)
+      return if @collection_type == 'groups'
```
Comment (hammady) [inline lib/rayyan-scrapers/mendeley_scraper.rb:33]: What does this do?

File: spec/mendeley_scraper_spec.rb (hunk 1 above)
Comment (hammady) [reply spec/mendeley_scraper_spec.rb]: Moving the discussion to the tested function.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (hammady) [reply README.md:131]: I don't fully agree with the implementation. It is hard to follow and does not fulfil all requirements. For example it returns 4 valid cases while they should only be 3. It contains repeated exceptions and repeated if statements. It raises an exception if there are arrays provided in the 2nd argument when the first one is empty/nil which is not stated in the requirement. I will try to write a pseudo code under that function.

File: lib/rayyan-scrapers/mendeley_scraper.rb [hunk 4]
```diff
@@ -129,14 +140,22 @@ def parse_search_results(page, extraction_fields = nil)
...
+
+        collection_ids.map { |id| "#{@base_url}/documents?group_id=#{id}" }
+      when nil, ''
+        return ["#{@base_url}/documents"] if collection_ids.all?(&:blank?)
+
+        raise 'UnknownCollectionType'
       else
-        raise "UnknownCollectionType"
+        raise 'UnknownCollectionType'
       end
```
Comment (hammady) [inline lib/rayyan-scrapers/mendeley_scraper.rb:158]: What about this? I think it is more readable and covers the requirements well.
```ruby
def search_urls(collection_type, collection_ids)
  if collection_type.blank?
    collection_type = ''
  else
    collection_ids&.compact!
    raise 'MissingCollectionIds' if collection_ids.blank?
  end

  case collection_type
  when ''
    return ["#{@base_url}/documents"]
  when 'folders'
    collection_ids.map { |id| "#{@base_url}/folders/#{id}/documents" }
  when 'groups'
    collection_ids.map { |id| "#{@base_url}/documents?group_id=#{id}" }
  else
    raise 'UnknownCollectionType'
  end
end
```

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (hammady) [reply README.md:131]: Check the other comment.

File: lib/rayyan-scrapers/mendeley_scraper.rb (hunk 3 above)
Comment (kohrVid) [reply lib/rayyan-scrapers/mendeley_scraper.rb:33]: One of your other comments mentioned that there was no need to iterate over the list pages for groups because that endpoint already returned a list of articles. I thought that because this method is called by the `ScraperBase#scrape` method, it would be easier to add an early return for groups. An alternative might have been to completely re-write the `#scrape` method in the `MendeleyScraper` so that this method isn't called at all for groups

Comment (hammady) [reply lib/rayyan-scrapers/mendeley_scraper.rb:33]: Yes, I remember that other comment, but I'm not sure if this is all what we need to do. You still need to iterate on list pages, but for groups, you should directly read articles from the list pages, without the need of fetching detail pages.

File: lib/rayyan-scrapers/mendeley_scraper.rb (hunk 4 above)
Comment (kohrVid) [reply lib/rayyan-scrapers/mendeley_scraper.rb:158]: To be honest, I think I was just struggling to understand the requirements in this case. I agree that the code in your example is much easier to follow...I'll update the PR in a bit

File: lib/rayyan-scrapers/mendeley_scraper.rb [hunk 5]
```diff
@@ -140,19 +140,18 @@ def parse_search_results(page, extraction_fields = nil)
...
       when 'groups'
-        return ["#{@base_url}/documents"] if collection_ids.all?(&:blank?)
-
-        collection_ids.map { |id| "#{@base_url}/documents?group_id=#{id}" }
-      when nil, ''
-        return ["#{@base_url}/documents"] if collection_ids.all?(&:blank?)
-
-        raise 'UnknownCollectionType'
+        collection_ids.map do |id|
+          query = "?group_id=#{id}" unless id.blank?
```
Comment (hammady) [inline lib/rayyan-scrapers/mendeley_scraper.rb:152]: Which case this check covers? The ungrouped documents are already covered by line 143.

File: lib/rayyan-scrapers/mendeley_scraper.rb [hunk 6]
```diff
@@ -140,19 +140,18 @@ def parse_search_results(page, extraction_fields = nil)
     private
 
     def search_urls(collection_type, collection_ids)
+      return ["#{@base_url}/documents"] if collection_type.blank?
+
       raise 'MissingCollectionIds' if collection_ids.blank?
```
Comment (hammady) [inline lib/rayyan-scrapers/mendeley_scraper.rb:145]: Missing the `&compact!` in my example code which removes nil ids.

Comment (kohrVid) [reply lib/rayyan-scrapers/mendeley_scraper.rb:145]: Oh right....I think I'm still covering the old requirement (where collection type is `groups` and the collections IDs contain empty strings) in that code and some of the tests. I'll fix that in a minute

Comment (kohrVid) [reply lib/rayyan-scrapers/mendeley_scraper.rb:145]: Although `compact!` doesn't remove empty strings. We'd need something like `compact_blank` which is a Rails method

Comment (hammady) [reply lib/rayyan-scrapers/mendeley_scraper.rb:145]: If you want to remove empty strings, you can `&.filter{|v| v.present?}` instead of compacting.

Comment (kohrVid) [reply lib/rayyan-scrapers/mendeley_scraper.rb]: Although the `filter` alias only appears in Ruby 2.6+ (I'm still using 2.3.0 to test locally). `select` seems to work fine in this case

File: lib/rayyan-scrapers/mendeley_scraper.rb [hunk 7]
```diff
@@ -142,16 +142,13 @@ def parse_search_results(page, extraction_fields = nil)
     def search_urls(collection_type, collection_ids)
       return ["#{@base_url}/documents"] if collection_type.blank?
 
-      raise 'MissingCollectionIds' if collection_ids.blank?
+      raise 'MissingCollectionIds' if collection_ids&.select { |id| id.present? }.blank?
```
Comment (hammady) [inline lib/rayyan-scrapers/mendeley_scraper.rb:145]: This does not remove nil/empty string ids from the collection, would `select!` solve this?

Comment (kohrVid) [reply lib/rayyan-scrapers/mendeley_scraper.rb:145]: Well `select` doesn't modify the `collection_ids` variable, no. It didn't seem necessary for this line but I guess you also don't want the empty strings to appear in the array further down in the case statement, right? If so, I can add the `!`

File: lib/rayyan-scrapers/mendeley_scraper.rb (hunk 3 above)
Comment (kohrVid) [reply lib/rayyan-scrapers/mendeley_scraper.rb:33]: I'm actually not sure how to handle this at the moment as it seems like part of the problem here is that I don't have a test explicitly checking that a `RayyanFormats::Target` instance is created for both collection types. When I check the output in the test application that I'm using, the list of articles always looks fine but that doesn't mean they're stored properly so adding the early return didn't reveal any obvious problems. But you're right in that we do need to iterate over pages and we don't need to make the extra API calls for groups...
I'll have to spend a little more time on this one

File: lib/rayyan-scrapers/mendeley_scraper.rb (hunk 7 above)
Comment (hammady) [reply lib/rayyan-scrapers/mendeley_scraper.rb:145]: Yes, that's right. My suggested implementation using `compact!` removes those intentionally so that they don't creep further down.

File: lib/rayyan-scrapers/mendeley_scraper.rb (hunk 3 above)
Comment (kohrVid) [reply lib/rayyan-scrapers/mendeley_scraper.rb]: Had a look at this again today. I think there were a few problems with the code in my earlier commit: for one, the line you highlighted wouldn't have worked because that instance variable was never set. It should be set now but I've had to refactor the `MendeleyScraper` quite a bit to to ensure that pages are still processed correctly and without the unnecessary API calls

File: lib/rayyan-scrapers/mendeley_scraper.rb (hunk 7 above)
Comment (kohrVid) [reply lib/rayyan-scrapers/mendeley_scraper.rb]: I had to call `.select!` on a separate line because [it also returns  `nil` if no changes are made](https://apidock.com/ruby/v2_5_5/Array/select%21). It's not very intuitive behaviour for a `!` method IMO but that's Ruby, I guess

File: lib/rayyan-scrapers/mendeley_scraper.rb [hunk 8]
```diff
@@ -94,17 +94,25 @@ def process_list_page(page, &block)
         id = item["id"]
         @logger.info "Got result with id #{id}"
         # get detailed info
-        process_detail_page(id, &block)
+
+        article = case @collection_type
```
Comment (hammady) [inline lib/rayyan-scrapers/mendeley_scraper.rb:98]: What about the 3rd case for the collection type?

**Comment by hammady** [comment]
@kohrVid Please update the PR description/README to document the third case where the collection type can be nil.

File: lib/rayyan-scrapers/mendeley_scraper.rb (hunk 8 above)
Comment (kohrVid) [reply lib/rayyan-scrapers/mendeley_scraper.rb:98]: Sorry - I've added that to the case statement now

File: README.md [hunk 9]
```diff
@@ -128,9 +128,23 @@ Similarly, grouped documents can be filtered using the `groups` collection type:
...
     end
 
-Note, if the `#scrape` method is called with the `groups` collection type and
-an empty list of collection IDs, the MendeleyScraper will return a list of
-ungrouped articles.
+Note, in the above examples, the `collection_ids` argument is required.
+
+Calling the `#scrape` method with a blank collection type (e.g., `nil` or `''`)
+will return an unfiltered list of documents. In this case, the `collection_ids`
+argument is optional:
```
Comment (hammady) [inline README.md:135]:
```suggestion
argument is ignored:
```

**Review by hammady — APPROVED** [review]

**Comment by hammady** [comment]
This looks great now, thanks @kohrVid!
